File: 2022/10.py
from utils import readInput
import statistics

class Compiler:
    def __init__(self, filename):
        self.lines = readInput(filename)
        self.wrong = ["(]", "(}", "(>", "[)", "[}", "[>",
                      "{)", "{]", "{>", "<)", "<]", "<}"]
        self.good = ["()", "[]", "{}", "<>"]
        self.counter = [0, 0, 0, 0]
        self.scores = []
    
    def check(self, l):
        found = -1
        while found != 0:
            found = 0
            for g in self.good:
                idx = l.find(g)
                if idx != -1:
                    l = l.replace(g, "", 1)
                    found += 1
                #print (found, l)

            for iw, w in enumerate(self.wrong):
                idx = l.find(w)
                if idx != -1:
                    if w[1] == ")":
                        self.counter[0] += 1
                    elif w[1] == "]":
                        self.counter[1] += 1
                    elif w[1] == "}":
                        self.counter[2] += 1
                    elif w[1] == ">":
                        self.counter[3] += 1
                    #print (w)
                    #print ("--------------")
                    return False
        return True

    def complete(self, l):
        found = -1
        while found != 0:
            found = 0
            for g in self.good:
                idx = l.find(g)
                if idx != -1:
                    l = l.replace(g, "", 1)
                    found += 1
                #print (found, l)
        score = 0
        for c in reversed(l):
            score *= 5
            if c == "(":
                score += 1
            elif c == "[":
                score += 2
            if c == "{":
                score += 3
            if c == "<":
                score += 4
        #print (score)
        self.scores.append(score)
        
    def scoringCorrupted(self):
        score = 0
        score += self.counter[0]*3
        score += self.counter[1]*57
        score += self.counter[2]*1197
        score += self.counter[3]*25137
        return score

c = Compiler("input_10.txt")
for l in c.lines:
    c.check(l)

print ("🎄 Part 1: {}".format(c.scoringCorrupted()))

for l in c.lines:
    if c.check(l):
        c.complete(l)
        
score = statistics.median(c.scores)
print ("🎄🎅 Part 2: {}".format(score))

File: 2022/10.jl
include("utils.jl")
using .AoCUtils: readInput

function loadInput()
    filename = "input_10.txt"
    lines = readInput(filename)
    adapters = parse.(Int, lines)
    push!(adapters, 0)
    push!(adapters, maximum(adapters) + 3)
    sort!(adapters)
    return adapters
end

function part1(adapters::Vector{Int})
    differences = Dict{Int, Int}(1=>0, 2=>0, 3=>0)
    for i in 2:length(adapters)
        differences[adapters[i] - adapters[i-1]] += 1
    end 
    println("🎄 Part 1: $(differences[1]*differences[3])")
end

function part2(adapters::Vector{Int})
    comb = Dict(0=>1, 1=>1, 2=>2, 3=>4, 4=>7)
    jolts = Int[]
    for i in 2:length(adapters)
        push!(jolts, adapters[i] - adapters[i-1])
    end 

    cons = Int[]
    c = 0
    for (i, j) in enumerate(jolts)
        if  j == 1
            c += 1
        else
            push!(cons, c)
            c = 0
        end
    end 

    tot = prod([comb[c] for c in cons])
    println("🎄🎅 Part 2: $(tot)")
end

println("⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄")
println("⛄        Day 10        ⛄")
println("⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄⛄")

inputs = loadInput()
@time part1(inputs)
@time part2(inputs)
